### PR TITLE
Abdullah JS3 week3

### PR DESCRIPTION
**Pull request by heartwhite**
JS3 week3 index.js the fetct function and async-await added

### CODE REVIEW

File: homework/App.js [hunk 1]
```diff
@@ -12,18 +12,35 @@ class App {
...
-    // 2. Make an initial XMLHttpRequest using Util.fetchJSON() to populate your <select> element
-
     const root = document.getElementById('root');
-
-    Util.createAndAppend('h1', root, { text: 'It works!' }); // TODO: replace with your own code
+    const header = Util.createAndAppend('header', root, { class: 'header' });
+    Util.createAndAppend('h3', header, { text: 'HYF Repositories' });
+    const selectMenu = Util.createAndAppend('select', header, {
+      class: 'repo-selector',
+      tabindex: 1,
```
Comment (remarcmij) [inline homework/App.js]: If you want to set the focus on the `<select>` element when the page loads,  setting the `tabindex` doesn't work. Use `autofocus` instead:

```js
 class: 'repo-selector',
 autofocus: true,
```

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#Attributes

File: homework/style.css [hunk 2]
```diff
@@ -1,3 +1,112 @@
...
+  padding: 0;
+  margin: 0;
+}
+.contributor-item {
+  border-bottom: solid 1px rgb(0, 0, 0, 12%);
+  padding: 16px;
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+  cursor: pointer;
```
Comment (remarcmij) [inline homework/style.css]: To get rid of the underlines in the list items you can add:

```css
text-decoration: none;
```

And you can remove `cursor: pointer` as that is the default for an anchor tag.

File: homework/index.js [hunk 3]
```diff
@@ -30,17 +15,118 @@
...
+      const root = document.getElementById('root');
+      createAndAppend('div', root, { text: err.message, class: 'alert-error' });
+    }
+  }
+
+  function mainHtmlConstructor(repos) {
+    makeHeader(repos);
+    const root = document.getElementById('root');
+    const container = createAndAppend('div', root, { id: 'container' });
+    createAndAppend('div', container, { class: 'leftDiv whiteframe' });
```
[the remark this answers is not in the document]
Comment (heartwhite) [reply homework/index.js]: my class names are converted to kebab case.

File: homework/index.js [hunk 4]
```diff
@@ -30,17 +15,118 @@
...
+    const selectMenu = document.querySelector('select');
+    makeHtmlOfContents(repos[selectMenu.selectedIndex]);
+    selectMenu.onchange = () => {
+      makeHtmlOfContents(repos[selectMenu.selectedIndex]);
+    };
+  }
+
+  async function main(url) {
+    try {
+      const responseOfHyfRepos = await fetch(url);
```
[the remark this answers is not in the document]
Comment (heartwhite) [reply homework/index.js]: yes. I handled error by using response.ok inside try area.
I read the documentation you sent. I hope I understood that

File: homework/index.js [hunk 5]
```diff
@@ -30,17 +15,118 @@
...
+      const repositories = await responseOfHyfRepos.json();
+      mainHtmlConstructor(
+        repositories.sort((one, two) =>
+          one.name.toLowerCase().localeCompare(two.name.toLowerCase()),
```
[the remark this answers is not in the document]
Comment (heartwhite) [reply homework/index.js]: I removed lowercase thing. it works

File: homework/App.js (hunk 1 above)
Comment (heartwhite) [reply homework/App.js]: I removed lowercase thing. it works

File: homework/App.js [hunk 6]
```diff
@@ -12,18 +12,35 @@ class App {
...
 
     try {
       const repos = await Util.fetchJSON(url);
       this.repos = repos.map(repo => new Repository(repo));
-      // TODO: add your own code here
+      this.repos
+        .sort((one, two) =>
+          one
+            .name()
+            .toLowerCase()
```
[the remark this answers is not in the document]
Comment (heartwhite) [reply homework/App.js]: I removed lowercase thing. it works

File: homework/style.css (hunk 2 above)
Comment (heartwhite) [reply homework/style.css]: yes. I applied text-decoration, removed th cursor thing.

File: homework/index.js [hunk 7]
```diff
@@ -30,17 +15,118 @@
...
+    createAndAppend('td', trForks, { text: 'Forks:' });
+    createAndAppend('td', trForks, { text: repo.forks_count });
+    const trUpdate = createAndAppend('tr', tBody);
+    createAndAppend('td', trUpdate, { text: 'Updated:' });
+    createAndAppend('td', trUpdate, {
+      text: new Date(repo.updated_at).toLocaleString(),
     });
   }
 
+  function makeContributorsList(repos) {
```
[the remark this answers is not in the document]
Comment (heartwhite) [reply homework/index.js]: oh! what I were thinking. ...
I changed to contributors

File: homework/index.js [hunk 8]
```diff
@@ -30,15 +15,115 @@
...
+    makeHtmlOfContents(repos[selectMenu.selectedIndex]);
+    selectMenu.onchange = () => {
+      makeHtmlOfContents(repos[selectMenu.selectedIndex]);
+    };
+  }
+
+  async function main(url) {
+    try {
+      const responseOfHyfRepos = await fetch(url);
+      const repositories = await responseOfHyfRepos.json();
```
Comment (remarcmij) [inline homework/index.js]: This must be done **_after_** you have checked for errors: if there is an error there is no JSON to convert.

File: homework/index.js [hunk 9]
```diff
@@ -30,15 +15,115 @@
...
+      if (!responseOfHyfRepos.ok) {
+        throw Error(responseOfHyfRepos.statusText);
```
Comment (remarcmij) [inline homework/index.js]: You must always prefix `Error` with `new` (although JavaScript may be forgiving if you forget it):

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error

File: homework/App.js [hunk 10]
```diff
@@ -12,18 +12,30 @@ class App {
...
+      autofocus: true,
+    });
+    Util.createAndAppend('div', root, { id: 'container' });
 
     try {
       const repos = await Util.fetchJSON(url);
       this.repos = repos.map(repo => new Repository(repo));
-      // TODO: add your own code here
+      this.repos
+        .sort((one, two) => one.name().localeCompare(two.name().toLowerCase()))
```
Comment (remarcmij) [inline homework/App.js]: There is still a `.toLowerCase()` in here, which shouldn't be there.

**Pull request closed without merging**